File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/RequireSingleTopLevelKotlinDeclarationRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.stream.Stream;

/**
 * Rule that requires Kotlin files to have exactly one top-level declaration.
 */
public enum RequireSingleTopLevelKotlinDeclarationRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireSingleTopLevelKotlinDeclaration";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        List<String> directories = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("directories"));
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return directories.stream()
                .flatMap(dir -> {
                    try {
                        Path dirPath = root.resolve(dir);
                        return HarnessCheckHelper.safeFileOrWalk(root, dirPath).stream()
                                .filter(f -> f.getFileName().toString().endsWith(".kt"))
                                .flatMap(file -> validateKotlinDeclarations(root, file, severity).stream());
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateKotlinDeclarations(Path root, Path file, String severity) throws MojoExecutionException {
        String text = HarnessCheckHelper.readFile(root, file);
        int count = 0;
        for (String line : text.split("\n")) {
            String trimmed = line.trim();
            if (trimmed.startsWith("class ") || trimmed.startsWith("fun ") || trimmed.startsWith("interface ") ||
                trimmed.startsWith("object ") || trimmed.startsWith("enum class ")) {
                count++;
            }
        }
        return count != 1
                ? List.of(new Finding(severity, CATEGORY, "Kotlin file must have exactly 1 top-level declaration: " + root.relativize(file)))
                : List.of();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireSingleTopLevelKotlinDeclarationRule.kt
package ai.harness.gradle.rules

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.exists
import kotlin.io.path.extension
import kotlin.io.path.relativeTo
import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessCheckRule
import ai.harness.gradle.HarnessPsiResults

/**
 * Rule that requires Kotlin files to have exactly one top-level declaration of allowed type.
 */
object RequireSingleTopLevelKotlinDeclarationRule : HarnessCheckRule {
	/**
	 * PSI analysis result for top-level declarations in a Kotlin source file.
	 */
	@Serializable
	data class Result(
		val file: String,
		val count: Int,
		val firstKind: String,
	)
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireSingleTopLevelKotlinDeclaration"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> {
		val category = "requireSingleTopLevelKotlinDeclaration"
		val severity = HarnessCheck.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject
		val parametersObj = catObj?.get("parameters")?.jsonObject
		val messagesObj = catObj?.get("messages")?.jsonObject
		val sourceRootsPerStack = parametersObj?.get("sourceRootsPerStack")?.jsonObject
		val extensionsPerStack = parametersObj?.get("extensionsPerStack")?.jsonObject
		return if (catObj == null || parametersObj == null || messagesObj == null || sourceRootsPerStack == null || extensionsPerStack == null) {
			emptyList()
		} else {
			val kotlinDirs = HarnessCheck.stringArrayFrom(sourceRootsPerStack, "kotlin")
			val kotlinExts = HarnessCheck.stringArrayFrom(extensionsPerStack, "kotlin")
			val allowedDeclarations = HarnessCheck.stringArrayFrom(parametersObj, "allowedDeclarations")
			val results = psiResults?.topLevelDeclarations ?: emptyList()
			kotlinDirs.flatMap { dirPattern ->
				val dir = root / dirPattern
				if (!dir.exists()) {
					emptyList()
				} else {
					val (files, _) = HarnessCheck.walkSafe(root, dir)
					files.filter { file ->
						file.extension in kotlinExts
					}.mapNotNull { file ->
						val info = results.find { it.file == file.name }
						when {
							info == null -> null
							info.count != 1 -> {
								val msg = HarnessCheck.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "${file.relativeTo(root)}: file must have single top-level declaration, found ${info.count}"
								Finding(severity, category, msg)
							}
							else -> {
								val declType = info.firstKind
								if (declType != "unknown" && declType !in allowedDeclarations) {
									val msg = HarnessCheck.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "${file.relativeTo(root)}: top-level $declType is not allowed"
									Finding(severity, category, msg)
								} else {
									null
								}
							}
						}
					}
				}
			}
		}
	}
}
